src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst

```diff
@@ -0,0 +1,211 @@
+open DY.Lib.Event.Typed
+open DY.Lib.State.PrivateKeys
+open DY.Lib.State.Tagged
+open DY.Lib.State.Typed
+
+open DY.Lib.Communication.Data
+open DY.Lib.Communication.Core
+open DY.Lib.Communication.RequestResponse
+open DY.Lib.Communication.Core.Invariants
+
+#set-options "--fuel 0 --ifuel 0 --z3cliopt 'smt.qi.eager_threshold=100'"
+
+(*** AEAD Predicate ***)
+
+#push-options "--ifuel 1"
+val aead_crypto_predicate_communication_layer: {|cusages:crypto_usages|} -> aead_crypto_predicate
+let aead_crypto_predicate_communication_layer #cusages = {
+  pred = (fun tr key_usage key nonce msg ad ->
+      (match parse authenticated_data ad with
+      | None -> False
+      | Some {server} -> (
+        event_triggered tr server (CommServerSendResponse server msg)
+      )
+    )
+  );
+  pred_later = (fun tr1 tr2 key_usage key nonce msg ad -> (
+    parse_wf_lemma authenticated_data (bytes_well_formed tr1) ad;
+    ()
+  ))
+}
+#pop-options
+
+val aead_crypto_predicates_communication_layer_and_tag:
+  {|cusages:crypto_usages|} ->
+  (string & aead_crypto_predicate)
+let aead_crypto_predicates_communication_layer_and_tag #cusages =
+  (comm_layer_aead_tag, aead_crypto_predicate_communication_layer)
+
+val has_communication_layer_reqres_crypto_predicates:
+  {|cinvs:crypto_invariants|} ->
+  prop
+let has_communication_layer_reqres_crypto_predicates #cinvs =
+  // Fix for the get_label function in the model code
+  cinvs.usages == default_crypto_usages /\
+  has_pke_predicate pke_crypto_predicates_communication_layer_and_tag /\
+  has_sign_predicate sign_crypto_predicates_communication_layer_and_tag /\
+  has_aead_predicate aead_crypto_predicates_communication_layer_and_tag
+
+
+(*** State Predicates ***)
+
+#push-options "--ifuel 2 --z3rlimit 25"
+let state_predicates_communication_layer {|crypto_invariants|}: local_state_predicate communication_states = {
+  pred = (fun tr prin sess_id st -> 
+    match st with
+    | ClientSendRequest {server; payload; key} -> (
+      let client = prin in
+      is_knowable_by (comm_label client server) tr payload /\
+      is_secret (comm_label client server) tr key /\
+      key `has_usage tr` (AeadKey comm_layer_aead_tag empty)
+    )
+    | ServerReceiveRequest {payload; key} -> (
+      let server = prin in
+      is_knowable_by (principal_label server) tr key /\
+      is_knowable_by (get_label tr key) tr payload /\
+      (
+        key `has_usage tr` (AeadKey comm_layer_aead_tag empty) \/
+        is_publishable tr key
+      )
+    )
+    | ClientReceiveResponse {server; payload; key} -> (
+      let client = prin in
+      is_knowable_by (comm_label client server) tr payload /\
+      is_secret (comm_label client server) tr key
+    )
+  );
+  pred_later = (fun tr1 tr2 prin sess_id state -> ());
+  pred_knowable = (fun tr prin sess_id state -> ());
+}
+#pop-options
+
+val state_predicates_communication_layer_and_tag: 
+  {|crypto_invariants|} ->
+  dtuple2 string local_bytes_state_predicate
```

Comment:
There doesn't seem to be any dependency going on here --- is there a specific reason to use dtuple2 and not tuple2/&? My impression is that non-dependent tuples should be better for performance (though probably not substantially so in the case where there is no dependency).

Reply:
We need `dtuple2` here because the function `val mk_state_pred: {|crypto_invariants|} -> list (dtuple2 string local_bytes_state_predicate) -> state_predicate` uses this type.